org.eclipse.jdt.launching/launching/org/eclipse/jdt/launching/JavaRuntime.java

```diff
@@ -3449,6 +3452,97 @@ public static VMStandin createVMFromDefinitionFile(File eeFile, String name, Str
+	}
+
+	/**
+	 * Returns the names of all packages provided by the given {@link IVMInstall vmInstall}, optionally restricted to the given
+	 * {@link IExecutionEnvironment execution-environment}.
+	 * <p>
+	 * The returned collection contains only distinct elements, sorted in alphabetical order.<br>
+	 * If the given VM-Install is {@link #isModularJava(IVMInstall) non modular}, the execution-environment is ignored.
+	 * </p>
+	 *
+	 * @param vm
+	 *            the vm-install to query
+	 * @param release
+	 *            the release to restrict the packages for, ignored for non-modular VMs
+	 * @return the distinct and alphabetically sorted collection of names of all packages in the VM
+	 * @throws CoreException
+	 *             if an error occurs or a modular VM is query for packages of Java-8 or lower
+	 * @since 3.21
+	 */
+	@SuppressWarnings("restriction")
+	public Collection<String> providedVMRuntimePackages(IVMInstall vm, String release) throws CoreException {
+		try {
+			if (vm instanceof AbstractVMInstall vmInstall) {
+				return vmInstall.systemPackagesCache.computeIfAbsent(release, r -> {
+					try {
+						return querySystemPackages(vmInstall, r);
+					} catch (CoreException e) {
+						throw new org.eclipse.jdt.internal.core.search.matching.MatchLocator.WrappedCoreException(e);
+					}
+				});
+			}
+		} catch (org.eclipse.jdt.internal.core.search.matching.MatchLocator.WrappedCoreException e) {
+			throw e.coreException;
+		}
+		return querySystemPackages(vm, release);
+	}
+
+	@SuppressWarnings("restriction")
+	private static final String JRT_PATH = "lib/" + org.eclipse.jdt.internal.compiler.util.JRTUtil.JRT_FS_JAR; //$NON-NLS-1$
+
+	@SuppressWarnings("restriction")
+	private Collection<String> querySystemPackages(IVMInstall vm, String release) throws CoreException {
+		if (!JavaRuntime.isModularJava(vm)) {
+			for (LibraryLocation libLocation : JavaRuntime.getLibraryLocations(vm)) {
+				IPath path = libLocation.getSystemLibraryPath();
+				if (path != null) {
+					try (ZipFile zip = new ZipFile(path.toFile())) {
+						return zip.stream().filter(e -> !e.isDirectory()).map(ZipEntry::getName).filter(n -> !n.startsWith("META-INF/")).map(n -> { //$NON-NLS-1$
+							int i = n.lastIndexOf('/');
+							return n.substring(0, i);
+						}).distinct().map(n -> n.replace('/', '.')).sorted().toList();
+					} catch (IOException e) {
+						throw new CoreException(Status.error("Failed to read packages in JVM library for " + vm + ", at " + path, e)); //$NON-NLS-1$//$NON-NLS-2$
+					}
+				}
+			}
+		}
+		if (JavaCore.compareJavaVersions(release, JavaCore.VERSION_1_8) <= 0) {
+			throw new UnsupportedOperationException("Cannot query a modular VM (JavaSE-9 or higher) for packages of JavaSE-1.8 and lower"); //$NON-NLS-1$
+		}
+		String path = new File(vm.getInstallLocation(), JRT_PATH).toString();
+		var jrt = org.eclipse.jdt.internal.core.builder.ClasspathLocation.forJrtSystem(path, null, null, release);
```

Comment:
i don't like the use of jdt internals in debug

Reply:
Me neither, but I don't have a better idea. If you have one, please let me know.
Besides that there are already many uses of jdt internals in this class and I think it is better done here than in PDE or somewhere else even further away from jdt-core.